retrieval/db/ingest.py

```diff
@@ -0,0 +1,147 @@
+import asyncio
+import logging
+import os
+
+from datetime import datetime
+from typing import Dict
+from typing import List
+
+import lancedb
+
+from dotenv import load_dotenv
+from dsp_nesta_brain import logger
+from langchain.docstore.document import Document as LangchainDocument
+from langchain.text_splitter import CharacterTextSplitter
+from openai import AsyncOpenAI
+from retrieval.db.schema import Chunk
+from retrieval.db.schema import Document as LanceDocument
+from scraping.scrape import search_query_to_scraped_data
+
+
+load_dotenv()
+
+os.environ["OPENAI_API_KEY"] = os.getenv("OPENAI_API_KEY")
+
+db = lancedb.connect("retrieval/db/ccid_demo_db")
+document_table = db.open_table("document")
+chunk_table = db.open_table("chunk")
+
+
+def already_in_db(location: str) -> bool:
+    """Determine whether chunks from a source document have already been added to the database"""
+    results = chunk_table.search().where(f'source.location = "{location}"').to_list()
+    return bool(results)
+
+
+def ingest(documents: List[LangchainDocument]) -> None:
+    """
+    Split documents into chunks, derive embeddings for the chunks
+    and insert Document and Chunk data (including embeddings) into the database
+    """  # noqa
+
+    async def to_Chunk(chunk: LangchainDocument) -> Chunk:
+        # intentionally not using the neater syntax documented by lanceDB which automatically calculates embeddings vectors
+        # using model.VectorField() specified in the schema.
+        # This is because I had issues getting the nested schema to work with this method.
+        async_client = AsyncOpenAI(api_key=os.getenv("OPENAI_API_KEY"))
+        source = [doc for doc in lance_documents if doc.location == chunk.metadata["location"]][0]
+        result = await async_client.embeddings.create(model="text-embedding-3-small", input=chunk.page_content)
```

Comment:
I haven't compared embeddings models so another might be better. If we want to compare them – or compare other design options – it would be a good idea to put a formal evaluation system in place.